データベース設計

Tsutomu-sasaki1 opened the pull request:
#What
ER図、README作成

#Why
機能実装時の指針

#ER図のurl
https://i.gyazo.com/c38e68510813c434dbab664a5368de46.png

よろしくお願いします。

Tsutomu-sasaki1 committed:
ER図、README

Tsutomu-sasaki1 committed:
user テーブル、unique: true を追加

masayaomoto commented on README.md:
```diff
@@ -1,24 +1,86 @@
+| Column           | Type   | Options                   |
+| ---------------- | ------ | ------------------------- |
+| email            | string | null: false               |
+| password         | string | null: false               |
```
```
| password         | string | null: false               |
```
こちらのカラムはdeviseのgemを使用しての実装になるため、
deviseが予め用意している```encrypted_password```というカラムを用意してあげましょう。

Tsutomu-sasaki1 replied:
修正しました。

masayaomoto commented on README.md:
```diff
@@ -1,24 +1,86 @@
+| name-kana-first  | string | null: false               |
+| name-kana-second | string | null: false               |
+| birthday         | date   | null: false               |
+| mail-address     | string | null: false, unique: true |
```
```| mail-address     | string | null: false, unique: true |```
emailに関してのカラムは9行目にて定義できておりますので、
17行目の記述は削除しておきましょう。

Tsutomu-sasaki1 replied:
修正しました。

masayaomoto commented on README.md:
```diff
@@ -1,24 +1,86 @@
+- has_many : histories
+- has_many : sell-items
+- has_many : comments
+- has_one : address
```
```
- has_one : address
```
ユーザーテーブルと配送先テーブルはアソシエーションを組む必要がございませんので削除しましょう。

Tsutomu-sasaki1 replied:
修正しました。

masayaomoto commented on README.md:
```diff
@@ -1,24 +1,86 @@
+| item-name        | string        | null: false                    |
+| category         | string        | null: false                    |
+| condition        | string        | null: false                    |
+| image            | ActiveStorage | null: false                    |
```
```
| image            | ActiveStorage | null: false                    |
```
こちらのカラムに関しまして、
今回の実装では画像投稿において「active_storage」を使用いたします。
active_storageにて画像データの管理を行いますのでこちらのカラムは不要となります。
削除しておきましょう！

Tsutomu-sasaki1 replied:
修正しました。

masayaomoto commented on README.md:
```diff
@@ -1,24 +1,86 @@
+| image            | ActiveStorage | null: false                    |
+| shipping-area    | string        | null: false                    |
+| date-of-shipment | string        | null: false                    |
+| price            | string        | null: false                    |
```
```
| price            | string        | null: false                    |
```
商品価格は数値で保存するためinteger型にしましょう！

Tsutomu-sasaki1 replied:
修正しました。

masayaomoto commented on README.md:
```diff
@@ -1,24 +1,86 @@
+| date-of-shipment | string        | null: false                    |
+| price            | string        | null: false                    |
+| shipping-charges | string        | null: false                    |
+| explanation      | string        | null: false                    |
```
```
| explanation      | string        | null: false                    |
```
こちらのカラムですが、
商品説明は文章になることも想定しtext型にしましょう！

Tsutomu-sasaki1 replied:
修正しました。

masayaomoto commented on README.md:
```diff
@@ -1,24 +1,86 @@
+| shipping-charges | string        | null: false                    |
+| explanation      | string        | null: false                    |
+| user             | references    | null: false, foreign_key: true |
 
```
商品の状態・配送料の負担・発送元の地域・発送までの日数・カテゴリー
に関してのカラムはactive_hashを用いての実装となります。
active_hashを用いたカラムはinteger型にしましょう！
また、カラム名には```_id```を付与しましょう！

Tsutomu-sasaki1 replied:
修正しました。

masayaomoto commented on README.md:
```diff
@@ -1,24 +1,86 @@
+### Association
 
-* Services (job queues, cache servers, search engines, etc.)
+- has_many : histories
```
```
- has_many : histories
```
「1つの商品は1つの購入情報を持っている」という設計にしたいため、
has_oneを使用して購入管理テーブルとのアソシエーションを組みましょう。

Tsutomu-sasaki1 replied:
修正しました。

masayaomoto commented on README.md:
```diff
@@ -1,24 +1,86 @@
+| Column           | Type          | Options                        |
+| ---------------- | ------------- | ------------------------------ |
+| postal-code      | string        | null: false                    |
+| prefectures      | string        | null: false                    |
```
```
| prefectures      | string        | null: false                    |
```
都道府県カラムはactive_hashを用いての実装となります。
active_hashを用いたカラムはinteger型にしましょう！
また、カラム名には```_id```を付与しましょう！

Tsutomu-sasaki1 replied:
修正しました。

masayaomoto commented on README.md:
```diff
@@ -1,24 +1,86 @@
+| prefectures      | string        | null: false                    |
+| city             | string        | null: false                    |
+| address          | string        | null: false                    |
+| building         | string        | null: false                    |
```
```
| building         | string        | null: false                    |
```
建物名に関しましては、
任意入力のデータとなりますので```null: false```は削除しましょう。

Tsutomu-sasaki1 replied:
修正しました。

masayaomoto commented on README.md:
```diff
@@ -1,24 +1,86 @@
+| address          | string        | null: false                    |
+| building         | string        | null: false                    |
+| phone-number     | string        | null: false                    |
+| user             | references    | null: false, foreign_key: true |
```
```| user             | references    | null: false, foreign_key: true |```
配送先テーブルにユーザーテーブルの外部キーは不要となりますので削除しましょう。

ユーザーテーブルの外部キーではなく購入管理テーブル（historiesテーブル）の外部キーを追加しましょう！

Tsutomu-sasaki1 replied:
修正しました。

masayaomoto commented on README.md:
```diff
@@ -1,24 +1,86 @@
+
+### Association
+
+- belongs_to : user
```
```
- belongs_to : user
```
今回の設計では「配送先住所は1つの購入情報に属している」という設計にしたいため、
ユーザーテーブルとのアソシエーションではなく、
購入管理テーブルとのアソシエーションを組みましょう。

Tsutomu-sasaki1 replied:
修正しました。

masayaomoto commented on README.md:
```diff
@@ -1,24 +1,86 @@
+### Association
+
+- belongs_to : sell-item
+- belongs_to : user
```
has_oneを使用して配送先テーブル（addressesテーブル）とのアソシエーションを組みましょう。

Tsutomu-sasaki1 replied:
修正しました。

※【historiesテーブル　association変更】の修正コミットでもアソシエーションを変更していますが、最終的にはこちらのコミットが最終となります。ややこしい事をしてしまい申し訳ありません。

masayaomoto reviewed (commented):
レビュー依頼ありがとうございます！
コメントしましたので、ご確認をお願いいたします！

Tsutomu-sasaki1 committed:
カラムencrypted_password追加

Tsutomu-sasaki1 committed:
９行目 mail-address 削除

Tsutomu-sasaki1 committed:
has_one : address　削除

Tsutomu-sasaki1 committed:
カラム ActiveStorage 削除

Tsutomu-sasaki1 committed:
priceカラム型　integerに修正

Tsutomu-sasaki1 committed:
explanationカラムの型 textに修正

Tsutomu-sasaki1 committed:
sell-item カラムはinteger型に修正

カラム名に_idを付与

Tsutomu-sasaki1 committed:
sell-itemテーブル has_oneに修正

Tsutomu-sasaki1 committed:
prefecturesカラムの型をintegerに修正

Tsutomu-sasaki1 committed:
buildingカラムのoption削除

Tsutomu-sasaki1 committed:
addressesテーブル 外部キー削除、追加

Tsutomu-sasaki1 committed:
historiesテーブル　association変更

Tsutomu-sasaki1 committed:
histories アソシエーション変更

【historiesテーブル　association変更】の修正コミットでもアソシエーションを変更していますが、最終的にはこちらのコミットが最終となります。ややこしい事をしてしまい申し訳ありません。

TakashiKozuma commented on README.md:
```diff
@@ -1,24 +1,84 @@
-* Ruby version
+| Column            | Type   | Options                   |
+| ----------------- | ------ | ------------------------- |
+| email             | string | null: false               |
```
```
| email             | string | null: false               |
```
下記のユーザー管理機能の実装条件に合うように、Optionsに記述を追記しましょう。
- メールアドレスが一意性であること

Tsutomu-sasaki1 replied:
修正しました。

TakashiKozuma commented on README.md:
```diff
@@ -1,24 +1,84 @@
+| Column            | Type   | Options                   |
+| ----------------- | ------ | ------------------------- |
+| email             | string | null: false               |
+| password          | string | null: false               |
```
```
| password          | string | null: false               |
```
パスワードに関するカラムは1つだけにしておきましょう。
実際には暗号化されていないpasswordは保存されず、暗号化されたencrypted_passwordのみが保存されます。

Tsutomu-sasaki1 replied:
修正しました。

TakashiKozuma commented on README.md:
```diff
@@ -1,24 +1,84 @@
+| name-first        | string | null: false               |
+| name-second       | string | null: false               |
+| name-kana-first   | string | null: false               |
+| name-kana-second  | string | null: false               |
```
```
| name-first        | string | null: false               |
| name-second       | string | null: false               |
| name-kana-first   | string | null: false               |
| name-kana-second  | string | null: false               |
```
カラム名にはハイフン"-"は用いないようにしましょう。
このまま登録しようとすると、エラーが出る場合があるためです。
他のテーブルに対しても同様の修正をお願いいたします。

Tsutomu-sasaki1 replied:
修正しました。

TakashiKozuma commented on README.md:
```diff
@@ -1,24 +1,84 @@
+| price_id            | integer       | null: false                    |
+| shipping-charges_id | string        | null: false                    |
+| explanation_id      | text          | null: false                    |
+| user_id             | references    | null: false, foreign_key: true |
```
```
| user_id             | references    | null: false, foreign_key: true |
```
references 型で記述する際は、カラム名の_idは不要となります。
もしカラム名で_idを使用する予定であれば、integer型を使用した方が良いでしょう。

Tsutomu-sasaki1 replied:
修正しました。

TakashiKozuma commented on README.md:
```diff
@@ -1,24 +1,84 @@
+| price_id            | integer       | null: false                    |
+| shipping-charges_id | string        | null: false                    |
+| explanation_id      | text          | null: false                    |
+| user_id             | references    | null: false, foreign_key: true |
```
```
| Column              | Type          | Options                        |
| ----------------    | ------------- | ------------------------------ |
| item-name_id        | string        | null: false                    |
| category_id         | integer       | null: false                    |
| condition_id        | integer       | null: false                    |
| shipping-area_id    | integer       | null: false                    |
| date-of-shipment_id | integer       | null: false                    |
| price_id            | integer       | null: false                    |
| shipping-charges_id | string        | null: false                    |
| explanation_id      | text          | null: false                    |
| user_id             | references    | null: false, foreign_key: true |
```
現状全てのカラム名に_idがついていますがアクティブハッシュなど一部必要なもの以外は_id外しておきましょう。それぞれのカラムの役割をわかりやすくするためです。

Tsutomu-sasaki1 replied:
修正しました。

Tsutomu-sasaki1 replied:
＃最新のER図です
https://i.gyazo.com/2c90f72107ec9b0eb32dfbdcb6138634.png

よろしくお願いします。

TakashiKozuma reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。アソシエーションも含め完成が近づいていますね！
コメントを追記いたしましたので、ご確認お願いします。
また、次回レビュー依頼時に最新状態のER図の添付もお願いいたします。

Tsutomu-sasaki1 committed:
emailカラム optionにunique追加

Tsutomu-sasaki1 committed:
passwordカラム削除

Tsutomu-sasaki1 committed:
各名称の"-"を"_"に修正

Tsutomu-sasaki1 committed:
カラム名の不要なidの削除

takahirohoriguchi commented on README.md:
```diff
@@ -1,24 +1,83 @@
+- belongs_to : sell_item
+- has_one : addresses
+
+## comments テーブル
+
+| Column         | Type       | Options                        |
+| -------------- | ---------- | ------------------------------ |
+| text           | string     | null: false                    |
+| user           | references | null: false, foreign_key: true |
+| sell_item      | references | null: false, foreign_key: true |
+
+### Association
+
+- belongs_to : sell_item
+- belongs_to : user
```
## comments テーブルは今回フリマの必須機能では使用しないため、削除しましょう！

Tsutomu-sasaki1 replied:
修正しました。

takahirohoriguchi commented on README.md:
```diff
@@ -1,24 +1,83 @@
+
+- belongs_to : history
+
+## histories　(中間)テーブル
+
+| Column       | Type       | Options                        |
+| ------------ | ---------- | ------------------------------ |
+| addresses    | references | null: false, foreign_key: true |
+| sell_item    | references | null: false, foreign_key: true |
+
+### Association
+
+- belongs_to : sell_item
+- has_one : addresses
```
購入管理テーブルは、ユーザーと商品のidのみで管理することが可能です。したがって、それ以外のカラムは今回は不要です！

Tsutomu-sasaki1 replied:
修正しました。（RE図も）

Tsutomu-sasaki1 replied:
現段階でRE図にaddressesが残っている状態ですが最終的なpushで訂正しました。

takahirohoriguchi commented on README.md:
```diff
@@ -1,24 +1,83 @@
+| address          | string        | null: false                    |
+| building         | string        |                                |
+| phone_number     | string        | null: false                    |
+| histories        | references    | null: false, foreign_key: true |
+
```
```
| histories        | references    | null: false, foreign_key: true |
```
historiesから**history**に修正しましょう！

Tsutomu-sasaki1 replied:
修正しました（ER図も）

takahirohoriguchi commented on README.md:
```diff
@@ -1,24 +1,83 @@
+| condition_id        | integer       | null: false                    |
+| shipping_area_id    | integer       | null: false                    |
+| date_of_shipment_id | integer       | null: false                    |
+| price_id            | integer       | null: false                    |
+| shipping_charges    | string        | null: false                    |
```
```
| shipping_charges    | string        | null: false                    |
```
こちらはactive_hashを使用するため「単数系カラム名_id」にしましょう。またinteger型にも修正しましょう。
理由：「_id」がない場合、後にエラーになってしまう可能性が高くなるからです。

Tsutomu-sasaki1 replied:
修正しました（ER図も）

takahirohoriguchi commented on README.md:
```diff
@@ -1,24 +1,83 @@
-* Configuration
+- has_many : histories
+- has_many : sell_items
+- has_many : comments
 
```
```
has_many : comments
```
こちらのアソシエーションは不要のため削除しましょう！

Tsutomu-sasaki1 replied:
commentsカラム削除時に修正済み。

takahirohoriguchi commented on README.md:
```diff
@@ -1,24 +1,83 @@
-* Services (job queues, cache servers, search engines, etc.)
+- has_one : histories
+- belongs_to : user
+- has_many : comments
 
```
```
 has_many : comments
```
こちらのアソシエーションは不要のため削除しましょう！

Tsutomu-sasaki1 replied:
commentsテーブル削除時に修正済み。

takahirohoriguchi commented on README.md:
```diff
@@ -1,24 +1,83 @@
+- has_many : sell_items
+- has_many : comments
 
-* Database creation
+## sell_item テーブル
 
-* Database initialization
+| Column              | Type          | Options                        |
+| ----------------    | ------------- | ------------------------------ |
+| item_name           | string        | null: false                    |
+| category_id         | integer       | null: false                    |
+| condition_id        | integer       | null: false                    |
+| shipping_area_id    | integer       | null: false                    |
+| date_of_shipment_id | integer       | null: false                    |
+| price_id            | integer       | null: false                    |
+| shipping_charges    | string        | null: false                    |
+| explanation         | text          | null: false                    |
+| user                | references    | null: false, foreign_key: true |
 
-* How to run the test suite
+### Association
 
-* Services (job queues, cache servers, search engines, etc.)
+- has_one : histories
```
```
has_one : histories
```
こちらhas_oneを使用する場合、単数系の**history**に修正しましょう！

Tsutomu-sasaki1 replied:
修正しました。

takahirohoriguchi commented on README.md:
```diff
@@ -1,24 +1,83 @@
+- has_many : comments
 
-* Database creation
+## sell_item テーブル
 
```
```
## sell_item テーブル
```
テーブル名は複数形にする必要があるため、sell_itemsに修正しましょう！

Tsutomu-sasaki1 replied:
修正しました。

takahirohoriguchi commented on README.md:
```diff
@@ -1,24 +1,83 @@
+# テーブル設計
 
-Things you may want to cover:
+## user テーブル
```
```
## user テーブル
```
テーブル名は複数形にする必要があるため、usersに修正しましょう！

Tsutomu-sasaki1 replied:
修正しました。

Tsutomu-sasaki1 replied:
#ER図の更新版です
https://i.gyazo.com/515e9fa3bd87ef153aaeae7d82841a7d.png

よろしくお願いします。

takahirohoriguchi reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました！完成に近づいています！
9点コメントしましたので、ご確認をお願いいたします！
あと少しです！応援しています！

Tsutomu-sasaki1 committed:
Update hurima.dio

Tsutomu-sasaki1 committed:
comments テーブルの削除

Tsutomu-sasaki1 committed:
購入管理テーブル及びRE図の修正

Tsutomu-sasaki1 committed:
historiesからhistoryに修正（ER図も）

Tsutomu-sasaki1 committed:
shipping_chargesを単数系+"_id"に修正

Tsutomu-sasaki1 committed:
has_one : historiesを単数系に修正

Tsutomu-sasaki1 committed:
sell_item テーブル　を複数系に修正

Tsutomu-sasaki1 committed:
## user テーブルを複数系に修正

Tsutomu-sasaki1 committed:
prefecturesに"_id"を追加

yuri240 commented on README.md:
```diff
@@ -1,24 +1,68 @@
-* Database initialization
+| Column              | Type          | Options                        |
+| ----------------    | ------------- | ------------------------------ |
+| item_name           | string        | null: false                    |
```
` item_name`ですが、テーブル名からitemであることは第三者が見てもわかるので 
`name`にしていただけると良いかと思います！

補足： 仮にインスタンスとして使用した場合` @item.item_name `だと コード量が長くなってしまう。 → `@item.name`の方がオススメ

Tsutomu-sasaki1 replied:
修正しました。

yuri240 commented on README.md:
```diff
@@ -1,24 +1,68 @@
+| condition_id        | integer       | null: false                    |
+| shipping_area_id    | integer       | null: false                    |
+| date_of_shipment_id | integer       | null: false                    |
+| price_id            | integer       | null: false                    |
```
`price_id`ですが、カラム名を`price`にしておきましょう！

理由：こちらはアクティブハッシュのデータではないため、`_id`をつける必要がない

Tsutomu-sasaki1 replied:
修正しました。

yuri240 commented on README.md:
```diff
@@ -1,24 +1,68 @@
+| Column           | Type          | Options                        |
+| ---------------- | ------------- | ------------------------------ |
+| postal_code      | string        | null: false                    |
+| prefectures_id   | integer       | null: false                    |
```
```
| prefectures_id   | integer       | null: false                    |
```
Active_hashのデータは使い回すことができますので、
`sell_itemsテーブル`の都道府県カラムと同じものを使用しましょう。

Tsutomu-sasaki1 replied:
修正しました。

yuri240 commented on README.md:
```diff
@@ -1,24 +1,68 @@
+| user         | references | null: false, foreign_key: true |
+| sell_item    | references | null: false, foreign_key: true |
+
+### Association
```
購入履歴テーブルに、addresseとのアソシエーションを組みましょう！

理由：購入する際には、配送先住所を入力する必要があるため。

また、こちらのアソシエーションは1対1を表す`has_one`を用いましょう！

Tsutomu-sasaki1 replied:
修正しました。

Tsutomu-sasaki1 replied:
#更新版ER図のURLです
https://i.gyazo.com/4c16b1f88738536a87ea4b7e13e624eb.png

yuri240 reviewed (commented):
レビュー依頼ありがとうございます！
ER図と修正点確認させていただきました。

追加で４点コメントさせていただきましたので、ご確認お願いします。

もうあと一息かと思いますので頑張りましょう！

Tsutomu-sasaki1 committed:
item_nameをnameへと修正

Tsutomu-sasaki1 committed:
price-idからpriceへと修正

Tsutomu-sasaki1 committed:
historiesとaddressをhas_oneで結ぶ

Tsutomu-sasaki1 committed:
カラムshipping_area_idをprefectures_idへ修正

NaturalRadio commented on README.md:
```diff
@@ -1,24 +1,72 @@
+| ------------ | ---------- | ------------------------------ |
+| user         | references | null: false, foreign_key: true |
+| sell_item    | references | null: false, foreign_key: true |
+| address      | references | null: false, foreign_key: true |
```
```
| address      | references | null: false, foreign_key: true |
```
商品の購入管理はuserとitemのidのみで管理が可能です。
したがって、それ以外のカラムは今回は不要です！

Tsutomu-sasaki1 replied:
修正しました。

NaturalRadio commented on README.md:
```diff
@@ -1,24 +1,72 @@
+
+- belongs_to : sell_item
+- belongs_to : user
+- belongs_to : address
```
```
belongs_to : address
```
購入情報は１つの住所を持っている設計にしたいので、`has_one :address`で住所テーブルとアソシエーションを組みましょう！

Tsutomu-sasaki1 replied:
修正しました。

NaturalRadio commented on README.md:
```diff
@@ -1,24 +1,72 @@
+### Association
+
+- belongs_to : history
+- has_one : history
```
```
has_one : history
```
こちら不要な記述となりますので削除しておきましょう！

Tsutomu-sasaki1 replied:
「addresとhistoryのアソシエーション修正」で修正済み。

Tsutomu-sasaki1 replied:
#更新版ER図
https://i.gyazo.com/85785912f75b7725ba49e5206df38a83.png

NaturalRadio reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いします！

Tsutomu-sasaki1 committed:
購入管理テーブルからadress削除

Tsutomu-sasaki1 committed:
addresとhistoryのアソシエーション修正

yuri240 reviewed (commented):
修正していただきありがとうございます。

問題なく設計できていることが確認できましたので、
LGTMとさせていただきます！

こちらのブランチは忘れずにMergeを行っていただき、
ローカルのMasterにPullしておきましょう。

データベース設計の実装、大変お疲れ様でした。
引き続き、応援しております。

Tsutomu-sasaki1 commented:
皆さま、レビューありがとうございました！
引続きよろしくお願いします！！
